File: src/utils/analytics.py
"""Analytics event logging to Firestore."""

import logging
from datetime import datetime, timezone
from typing import Optional

from src.utils.firebase_admin import get_firestore_client

logger = logging.getLogger(__name__)


def log_query(
    user_id: Optional[str],
    session_id: Optional[str],
    question: str,
    latency_ms: int,
    chunks_used: int,
    sources_count: int,
    figures_count: int,
) -> None:
    """Log a query event to Firestore analytics/queries collection (best-effort)."""
    try:
        db = get_firestore_client()
        db.collection("analytics").document("queries").collection("events").add(
            {
                "userId": user_id or "anonymous",
                "sessionId": session_id or "",
                "questionPreview": question[:80],
                "latencyMs": latency_ms,
                "chunksUsed": chunks_used,
                "sourcesCount": sources_count,
                "figuresCount": figures_count,
                "isAnonymous": user_id is None,
                "timestamp": datetime.now(timezone.utc),
            }
        )
    except Exception as e:
        logger.warning(f"Failed to log query analytics: {e}")


def log_daily_metrics(latency_ms: int, is_error: bool = False) -> None:
    """Increment daily system metrics (best-effort)."""
    try:
        from google.cloud.firestore import Increment

        db = get_firestore_client()
        today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        doc_ref = (
            db.collection("system")
            .document("daily")
            .collection("metrics")
            .document(today)
        )
        update = {
            "totalQueries": Increment(1),
            "totalLatencyMs": Increment(latency_ms),
        }
        if is_error:
            update["errorCount"] = Increment(1)
        doc_ref.set(update, merge=True)
    except Exception as e:
        logger.warning(f"Failed to update daily metrics: {e}")

File: askchuck-frontend/src/lib/analytics.ts
import { db } from './firebase'
import { collection, addDoc, serverTimestamp } from 'firebase/firestore'

type EventType = 'citation_clicked' | 'pdf_opened' | 'session_started' | 'login' | 'signup' | 'query_completed'

export async function logEvent(
  userId: string | null,
  sessionId: string | null,
  eventType: EventType,
  metadata: Record<string, unknown> = {}
): Promise<void> {
  try {
    await addDoc(collection(db, 'analytics', 'events', 'log'), {
      userId: userId || 'anonymous',
      sessionId: sessionId || '',
      eventType,
      metadata,
      timestamp: serverTimestamp(),
    })
  } catch {
    // Best-effort, never throw
  }
}

export async function logQuery(
  userId: string | null,
  sessionId: string | null,
  question: string,
  latencyMs: number,
  sourcesCount: number,
  figuresCount: number
): Promise<void> {
  try {
    await addDoc(collection(db, 'analytics', 'queries', 'log'), {
      userId: userId || 'anonymous',
      sessionId: sessionId || '',
      questionPreview: question.slice(0, 80),
      latencyMs,
      sourcesCount,
      figuresCount,
      isAnonymous: userId === null,
      timestamp: serverTimestamp(),
    })
  } catch {
    // Best-effort, never throw
  }
}
